server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,39 +268,198 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
+        apiInfo = Pa_GetHostApiInfo(Pa_GetDeviceInfo(*knownDevice)->hostApi);
+
+        bool isInput;
+        if (matchingDeviceType == IOType::Input)
+            isInput = true;
+        else if (matchingDeviceType == IOType::Output)
+            isInput = false;
+
+        if (isInput) {
+            *matchingDevice = apiInfo->defaultInputDevice;
+        } else {
+            *matchingDevice = apiInfo->defaultOutputDevice;
         }
+        if (*matchingDevice != paNoDevice)
+            fprintf(stdout, "Selecting default %s %s device\n", apiInfo->name, (isInput ? "input" : "output"));
     }
 }
 
+// this function will select default PA devices if they are not defined
+// it will also try to check for some configuration problems
+// numIns, numOuts and sampleRate are only the requested values, can change later
+PaError SC_PortAudioDriver::CheckPaDevices(int* inDevice, int* outDevice, int numIns, int numOuts, double sampleRate) {
+    // make independent checks whether we are using only input, only output, or both input and outputTouched
+    if (numIns && !numOuts) {
+        // inputs only
+        *outDevice = paNoDevice;
+        // check for requested sample rate or select the default device
+        return CheckSinglePaDevice(inDevice, sampleRate, IOType::Input);
+    } else if (!numIns && numOuts) {
+        // outputs only
+        *inDevice = paNoDevice;
+        // check for requested sample rate or select the default device
+        return CheckSinglePaDevice(outDevice, sampleRate, IOType::Output);
+    } else if (numIns && numOuts) {
+        // inputs and outputs
+        // if one device is specified, let's try to open another one on matching api
+        // try matching input to output
+        SelectMatchingPaDevice(inDevice, outDevice, IOType::Input);
+        // then try matching output to input
+        SelectMatchingPaDevice(outDevice, inDevice, IOType::Output);
+        // check if devices are having mismatched API, but only if they are defined
+        if (*inDevice != paNoDevice && *outDevice != paNoDevice
+            && Pa_GetDeviceInfo(*inDevice)->hostApi != Pa_GetDeviceInfo(*outDevice)->hostApi) {
+            fprintf(stdout, "Requested devices %s and %s use different API. ", GetPaDeviceName(*inDevice).c_str(),
+                    GetPaDeviceName(*outDevice).c_str());
+            *outDevice = Pa_GetHostApiInfo(Pa_GetDeviceInfo(*inDevice)->hostApi)->defaultOutputDevice;
+            fprintf(stdout, "Setting output device to %s.\n", GetPaDeviceName(*outDevice).c_str());
+        }
+        // check for matching sampleRate or requested sample rate
+        if (*inDevice != paNoDevice && *outDevice != paNoDevice) {
+            PaStreamParameters in_parameters, out_parameters;
+            in_parameters = GetPaStreamParameters(inDevice, Pa_GetDeviceInfo(*inDevice)->maxInputChannels, 0);
+            out_parameters = GetPaStreamParameters(outDevice, Pa_GetDeviceInfo(*outDevice)->maxOutputChannels, 0);
```

Comment:
can declare these variables much more simply -- `const auto in_params = GetPaStreamParameters(...);` in both cases